### ISSUE TITLE
Color badges in the `Administration` views of Customers, Projects, Activities

### ISSUE THREAD

**Issue by sharky98** [kimai/kimai#4186]
### Describe the issue

The bug is related to the color badge shown on the right side of the name. There is different variations that I'll put inside this single issue. Feel free to split as needed.

** TL;DR **

The color shown on badge beside Customers, Projects, Activities is not consistent across views and varies based on various conditions when no color was selected (in other words the empty color was selected).

### (1) List view, `Name` column, Uneven behavior when leaving color choice empty

When creating or editing a Customer, Project or Activity and leaving the color field empty (confirmed by `"color": null` in the API), the behavior of the color used seems random. I have observed two behaviors so far.

1. Default color is set to `#2d3748`.
    * There seems to have a pattern to trigger this one. Customer with name `another brand new customer`, `fake`, `maroon first, empty second` and `test` get this code.
2. Default color is set to random value (`#009688`, `#dddddd`, `#ffdc00`, `#9c27b0` and few others).

For the other section, I will call the color selected by this unknown behavior as the `empty random color`.

### (2) Detail view, Details box uneven behavior

In the header of the detail box, for the main title of

- Customer, the color shown in the badge follows the `empty random color`.
- Project, the color shown in the badge is always `#929dab`.
- Activity, the color shown in the badge is always `#929dab`.

In the details of the detail box, 

- Customer, the color shown in the badge follows the `empty random color`.
- Project, the color shown in the badge follows two distinct behaviors.
    1. If the Customer is also empty, the Project badge color will be the `empty random color` of the project.
    2. If the Customer has a selected color, the Project badge color will use the Customer color.

### (3) Empty Project color does not always inherit Customer

When creating or editing a Project and leaving the color field empty (confirmed by `"color": null` in the API), the color of the badge shown seems to follow different rules.

* In the List view of Project, `Name` column, the colors shown is the `empty random color`.
* In the List view of Activities, in the `Project` column there is two distinct behaviors.
    1. If the Customer is also empty, the Project badge color will be the `empty random color`.
    2. If the Customer has a selected color, the Project badge color will use the Customer color.
* In the Detail view of the Customers, in the Project card there is two distinct behaviors.
    1. If the Customer is also empty, the Project badge color will be the `empty random color`.
    2. If the Customer has a selected color, the Project badge color will use the Customer color.

### (4) Empty Activity color does not always inherit Project

The behavior is similar to the Project described in section (3) above. When creating or editing an Activity and leaving the color field empty (confirmed by `"color": null` in the API), the color of the badge shown seems to follow different rules.

* In the List view of Activities, `Name` column, the colors shown is the `empty random color`.
* In the Detail view of Projects, in the Activities card there is three distinct behaviors.
    1. If the Project color is empty and the Customer is also empty, the Activity badge color will be the `empty random color`.
    2. If the Project color is empty, but the Customer is not empty, the Activity badge color will be the same as the Customer.
    3. If the Project color is not empty, the Activity badge color will be the same as the Project.

### Some ideas for solution

For the Customer, if the default behavior is to use a random color, I think it would be best for stability to save the randomly selected color as part of the entity.

For the Project and Activities, since the behavior is to inherit or use selected color, leaving an empty field would make sense. However, the algorithm to use a random color should not be applied to them (they either inherit or have a selected color).

### What documentation says

I've copied what the manual says here as future reference.

#### Customers

> If a project or an activity linked to the customer does not have its own color, the customers color will be used. The default color code is #d2d6de.

#### Projects

> If no color is applied, Kimai will fall back to the customers color and finally to the [default color](https://www.kimai.org/documentation/customer.html).

#### Activities

> If no color has been selected, the project color will be used. If that is not defined as well, Kimai will fall back to the customer color and finally to the [default color](https://www.kimai.org/documentation/customer.html).

### Searched issues

- #2398
- #1919

### Searched discussions
- #3446

### I already tried

- [X] I've read and searched [the documentation](https://www.kimai.org/documentation/).
- [X] I've searched for similar issues in this repository.
- [X] I've searched for similar issues in [the discussions](https://github.com/kimai/kimai/discussions).

### Kimai version

2.0.23

### How do you run Kimai?

Docker

### Which PHP version are you using?

8.1

**Comment by kevinpapst** [comment]
1. Thanks for your super detailed input!
2. The docs are outdated.
3. If no color is given, it is calculated by the name, so same name should lead to same color. 
4. The color should not change randomly. 
5. The default color is hopefully not used anymore, but from reading your issue, it seems it still is in some places.
6. Saving the color would prevent to apply a new color calculation. Not sure if I like that idea.

If anyone is interested to send a PR to fix this, please go ahead. 
I can help identifying the places where to look at.

**Comment by sharky98** [comment]
> 3. If no color is given, it is calculated by the name, so same name should lead to same color.
> 4. The color should not change randomly.

True, after some more check this seems to make sense regarding the behavior. I should then change my "bug" observation that it seems the calculation have a biais towards the color `#2d3748` (this is the only one that repeated in my limited manual testing, which is the reason I though it was a different case). Not sure if there is something to do about it though if its calculated.

> 6. Saving the color would prevent to apply a new color calculation. Not sure if I like that idea.

That is true it would prevent a new color calculation. In my opinion, I think this should be a desired behavior.

My reasoning is that, in most cases, an admin would **not** want to change all historic data made under customer `Company A` to a unrelated `Company B` just by changing the name. As such, I would assume most changes would occurs to correct typos, add some kind of tags[^1], modify customer legal reason, etc. Under that reasoning, I believe it should be good UX practice to keep a stable color even if the name changes after the initial creation using a random color[^2] to prevent user confusion[^3].

Here some tool I use that come to mind that have similar feature of random color. They all use similar feature that "lock" in place the color once generated on the first creation.
- GitHub tags
   - Required input field (required, check it is hex-color format)
   - Focus on field trigger modal with list of preset colors.
   - Button on the right side of field to generate a color (random? pre-made list? pseudo-random?).
- GitLab tags
   - Required input field (required, check it is hex-color format)
   - A list of preset colors is shown below.
   - Button on the right side of field to have a color picker (looks like the browser default color picker).
- Clockify item Badges (well, I'm not using it anymore, since I use Kimai now😄!)
   - The color is selected from a dropdown.
   - A different color is pre-selected from the preset list when creating a new item.
   - When using the dropdown, the preset list is shown and a link to use a color picker (which looks like a custom one).

I think a sensible solution UX-wise in Kimai would be to keep the same dropdown feature with few rules:
- When creating a customer with the "empty" choice, pick a color at random from the configured list in the settings.
- Save that color with the entity.
- If the user re-select the "empty" choice when editing, reselect a new random color from the configured list.

For Project and Activities it should be "empty" = inherit from parent entity, do not use random color. Should be the same as now (except for the issues raised in here).

Or (as a new feature), they could be made to add specifically "Inherit" as a choice, and leave the "empty" choice picking a color at random from the configured list. Leaving "empty" as the default choice.


[^1]: In my case I added `[HM]` to some customers name to flag to myself they are "Home" customers, since I use customer more like project groups than real customer.
[^2]: I am not an UX expert, far from that. However, I base my opinion from experience and other tools that use similar feature of generating colors; once the entity is created the color is saved with it and will not change (unless the user request it).
[^3]: Let's imagine that the admin change `Company A` (that generated a red badge) to `Company A inc.` (coming with a blue badge). Next morning, I would expect many users calling their helpdesk with question like "`Company A` disappeared from the list of customers!!"

**Comment by kevinpapst** [comment]
Thinking about it: 

- All this code just exists to make sure there is a color
- Inheritance was important when adding the feature, but I think that was just something I imagined would be useful. Today I am not sure if that still applies.

How to fix:

- Logic could be greatly reduced (improving runtime as well) by simply forcing a color to exist
- A migration could update all database fields which are currently null

But that has several drawbacks, especially for the users that are used to the current solution, some of which might prefer the inheritance.

If we only force the color on the customer, inheritance could be used to fix all these automatic calculations and fallback color.

**Comment by sharky98** [comment]
> If we only force the color on the customer, inheritance could be used to fix all these automatic calculations and fallback color.

In my opinion, this would be the ideal solution.

* Customer
  * Selected color
  * or, Forced color
* Project
  * Selected color
  * or, Inherited color
* Activity
  * Selected color
  * or, Inherited color

And since I created that issue, I started using tags, so I would follow the same kind of rules for tags as customers.

* Tags
  * Selected color
  * or, Forced color

As for the migration to limit breaking changes to end-users, I would imagine to change the database schema to have the color fields `NOT NULL` for customers and tags and get a migration script that use the current logic for color based on name to fill the `NULL` ones. This would be non-breaking changes for users.

As for the teamleads and administrators, There would be a small breaking change on how the color dropdown behave. The empty select could be changed to have the text "Random" in `Customer` and `Tags` (and it would select a random color from the allowed list) and "Inherit" in `Project` and `Activity` (by keeping the current behavior on `NULL`, inherit).
